rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md

```diff
@@ -0,0 +1,458 @@
+slippery concept without concrete examples (at least that's what I found when
+writing this document).
+
+```
+origin_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+  {id: 2, mature: true}, # no sensitive text
+  {id: 3, mature: false}, # has sensitive text
+  {id: 4, mature: true}, # has sensitive text
+]
+
+filtered_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+]
+
+origin_index_result_ids - filtered_index_result_ids
+# => [1, 2, 3, 4] - [1] = [2, 3, 4]
+```
+
+Note how the final resulting list of IDs includes results marked mature, but
+that do not have sensitive textual content (result 2). Unfortunately there is no
+way around this as if we were to exclude the results marked `mature` from the
+origin result IDs in the final set subtraction, we'd end up with the following
+situation:
+
+```
+origin_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+  {id: 2, mature: true}, # no sensitive text
+  {id: 3, mature: false}, # has sensitive text
+  {id: 4, mature: true}, # has sensitive text
+]
+
+filtered_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+]
+
+origin_index_less_mature_result_ids = [r for r in origin_index results if not r["mature"]]
+# => [1, 3]
+
+origin_index_less_mature_result_ids - filtered_index_result_ids
+# => [1, 3] - [1] = [3]
+```
+
+Notice how we still have a problem: we've excluded the result that has sensitive
+text _and_ is marked mature (result 4).
+
+It's tempting to think we could move the mature filtering to _after_ when we've
+derived the list of sensitive result IDs. However, the problem will remain the
+same: we will exclude results that are both marked mature and have sensitive
+textual content:
+
+```
+origin_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+  {id: 2, mature: true}, # no sensitive text
+  {id: 3, mature: false}, # has sensitive text
+  {id: 4, mature: true}, # has sensitive text
+]
+
+filtered_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+]
+
+sensitive_result_ids = origin_index_result_ids - filtered_index_result_ids
+# => [1, 2, 3, 4] - [1] = [2, 3, 4]
+
+sensitive_results = [r for r in origin_index_results if r["_id"] in sensitive_result_ids and not r["mature"]]
+# => [3]
+```
+
+Result 4 is still excluded because we can't tell in the final operation whether
+result 4 was not in the set of IDs from the filtered index _only_ because it is
+marked mature or if it also has sensitive textual content.
+
+The only way I can think of to solve this is to have an index solely of
+documents that _do_ have sensitive textual content and querying it to derive the
+final list of sensitive results without needing to rely on the diversely
+filtered index. However, this adds a big re-indexing burden (yet another
+additional index) that can be bypassed altogether if we just exclude mature
+results from the filtered index to begin with.
```

Comment:
Thank you for these examples! The problem is quite clear and your reasoning here makes sense to me.